src/stream.rs

```diff
@@ -771,11 +779,17 @@ impl SendBuf {
             out_len -= buf.len();
             out_off = buf.max_off();
 
-            out.fin = out.fin || buf.fin();
-
             out.data.extend_from_slice(&buf.data);
         }
 
+        // Ignore the `fin` flag set for the range buffer itself, and instead
+        // match the buffer's maximum offset against the known final size (if
+        // any). This way we can avoid queueuing empty buffers just so we can
+        // track the final size.
+        if self.fin_off == Some(out.max_off()) {
```

Comment:
Do we need to have `>=` instead of `==`?

Reply:
No, the maximum offset of the outbut buffer needs to match the final offset exactly, otherwise we end up setting `fin=true` for multiple STREAM frames.